File: examples/agent-client/test-agent.ts
/**
 * Basic tests for the BioVault Agent Client (Node.js/TypeScript)
 *
 * These tests require a running BioVault Desktop instance with the
 * WebSocket bridge enabled.
 *
 * Run with:
 *   npx ts-node test-agent.ts
 */

import { BioVaultAgent, BioVaultAgentError } from './biovault-agent'

interface TestResult {
	name: string
	passed: boolean
	error?: string
	output?: string
}

const results: TestResult[] = []

async function runTest(
	name: string,
	testFn: (agent: BioVaultAgent) => Promise<void>,
): Promise<void> {
	console.log(`\nTest: ${name}`)
	const agent = new BioVaultAgent()

	try {
		await agent.connect()
		await testFn(agent)
		results.push({ name, passed: true })
		console.log('  Result: PASSED')
	} catch (err) {
		const error = err instanceof Error ? err.message : String(err)
		results.push({ name, passed: false, error })
		console.log(`  Result: FAILED - ${error}`)
	} finally {
		await agent.disconnect()
	}
}

// -------------------------------------------------------------------------
// Tests
// -------------------------------------------------------------------------

async function testConnection(agent: BioVaultAgent): Promise<void> {
	if (!agent.connected) {
		throw new Error('Not connected')
	}
	console.log('  Connection: OK')
}

async function testDiscover(agent: BioVaultAgent): Promise<void> {
	const result = await agent.discover()
	if (!result.version) throw new Error('Missing version')
	if (!result.name) throw new Error('Missing name')
	if (result.name !== 'BioVault Desktop Agent API') {
		throw new Error(`Unexpected name: ${result.name}`)
	}
	console.log(`  API Version: ${result.version}`)
}

async function testListCommands(agent: BioVaultAgent): Promise<void> {
	const result = await agent.listCommands()
	if (!result.version) throw new Error('Missing version')
	if (!Array.isArray(result.commands)) throw new Error('Missing commands array')
	if (result.commands.length < 50) {
		throw new Error(`Expected 50+ commands, got ${result.commands.length}`)
	}
	console.log(`  Commands: ${result.commands.length}`)
}

async function testGetAppVersion(agent: BioVaultAgent): Promise<void> {
	const version = await agent.getAppVersion()
	if (typeof version !== 'string') throw new Error('Version not a string')
	if (version.length === 0) throw new Error('Version is empty')
	console.log(`  App Version: ${version}`)
}

async function testIsDevMode(agent: BioVaultAgent): Promise<void> {
	const isDevMode = await agent.isDevMode()
	if (typeof isDevMode !== 'boolean') throw new Error('Not a boolean')
	console.log(`  Dev Mode: ${isDevMode}`)
}

async function testCheckIsOnboarded(agent: BioVaultAgent): Promise<void> {
	const isOnboarded = await agent.checkIsOnboarded()
	if (typeof isOnboarded !== 'boolean') throw new Error('Not a boolean')
	console.log(`  Onboarded: ${isOnboarded}`)
}

async function testGetSyftboxState(agent: BioVaultAgent): Promise<void> {
	const state = await agent.getSyftboxState()
	if (typeof state !== 'object') throw new Error('Not an object')
	if (!('running' in state)) throw new Error('Missing running field')
	console.log(`  SyftBox Running: ${state.running}`)
}

async function testCheckDependencies(agent: BioVaultAgent): Promise<void> {
	const deps = await agent.checkDependencies()
	if (!Array.isArray(deps)) throw new Error('Not an array')
	console.log(`  Dependencies: ${deps.length} found`)
	for (const dep of deps.slice(0, 3)) {
		const d = dep as { name: string; installed: boolean }
		const status = d.installed ? 'installed' : 'missing'
		console.log(`    - ${d.name}: ${status}`)
	}
}

async function testGetProjects(agent: BioVaultAgent): Promise<void> {
	const projects = await agent.getProjects()
	if (!Array.isArray(projects)) throw new Error('Not an array')
	console.log(`  Projects: ${projects.length} found`)
}

async function testSqlListTables(agent: BioVaultAgent): Promise<void> {
	const tables = await agent.sqlListTables()
	if (!Array.isArray(tables)) throw new Error('Not an array')
	console.log(`  Database Tables: ${tables.length}`)
	for (const table of tables.slice(0, 5)) {
		console.log(`    - ${table}`)
	}
}

async function testAuditLog(agent: BioVaultAgent): Promise<void> {
	const log = await agent.getAuditLog(5)
	if (!Array.isArray(log)) throw new Error('Not an array')
	console.log(`  Audit Log Entries: ${log.length}`)

	if (log.length > 0) {
		const entry = log[log.length - 1] as { timestamp?: string; cmd?: string }
		if (!entry.timestamp) throw new Error('Missing timestamp')
		if (!entry.cmd) throw new Error('Missing cmd')
	}
}

async function testInvalidCommand(agent: BioVaultAgent): Promise<void> {
	try {
		await agent.invoke('this_command_does_not_exist')
		throw new Error('Should have raised an error')
	} catch (err) {
		if (err instanceof BioVaultAgentError) {
			if (!err.message.includes('Unhandled command')) {
				throw new Error(`Unexpected error message: ${err.message}`)
			}
			console.log('  Invalid command handling: OK')
		} else {
			throw err
		}
	}
}

async function testAuthWithWrongToken(): Promise<void> {
	// Only run if token auth is configured
	if (!process.env.AGENT_BRIDGE_TOKEN) {
		console.log('  Skipping (no token configured)')
		return
	}

	const agent = new BioVaultAgent({ token: 'wrong-token' })
	try {
		await agent.connect()
		await agent.getAppVersion()
		throw new Error('Should have been rejected')
	} catch (err) {
		if (err instanceof BioVaultAgentError) {
			if (!err.message.includes('Authentication failed')) {
				throw new Error(`Unexpected error: ${err.message}`)
			}
			console.log('  Auth rejection: OK')
		} else {
			throw err
		}
	} finally {
		await agent.disconnect()
	}
}

// -------------------------------------------------------------------------
// Main
// -------------------------------------------------------------------------

async function runAllTests(): Promise<boolean> {
	console.log('\nBioVault Agent Client Tests (Node.js)')
	console.log('='.repeat(50))

	await runTest('Connection', testConnection)
	await runTest('API Discovery', testDiscover)
	await runTest('List Commands', testListCommands)
	await runTest('Get App Version', testGetAppVersion)
	await runTest('Dev Mode Check', testIsDevMode)
	await runTest('Onboarding Status', testCheckIsOnboarded)
	await runTest('SyftBox State', testGetSyftboxState)
	await runTest('Dependencies', testCheckDependencies)
	await runTest('Projects', testGetProjects)
	await runTest('SQL Tables', testSqlListTables)
	await runTest('Audit Log', testAuditLog)
	await runTest('Invalid Command', testInvalidCommand)

	// Special test that creates its own agent
	console.log('\nTest: Auth Rejection')
	try {
		await testAuthWithWrongToken()
		results.push({ name: 'Auth Rejection', passed: true })
		console.log('  Result: PASSED')
	} catch (err) {
		const error = err instanceof Error ? err.message : String(err)
		results.push({ name: 'Auth Rejection', passed: false, error })
		console.log(`  Result: FAILED - ${error}`)
	}

	// Summary
	const passed = results.filter((r) => r.passed).length
	const failed = results.filter((r) => !r.passed).length

	console.log('\n' + '='.repeat(50))
	console.log(`Results: ${passed} passed, ${failed} failed`)
	console.log('='.repeat(50))

	return failed === 0
}

// Run tests
runAllTests()
	.then((success) => process.exit(success ? 0 : 1))
	.catch((err) => {
		console.error('Test runner error:', err)
		process.exit(1)
	})

File: examples/agent-client/test_agent.py
#!/usr/bin/env python3
"""
Basic tests for the BioVault Agent Client.

These tests require a running BioVault Desktop instance with the
WebSocket bridge enabled.

Run with:
    pytest test_agent.py -v

Or without pytest:
    python test_agent.py
"""

import asyncio
import os
import sys

# Add parent directory to path for imports
sys.path.insert(0, os.path.dirname(__file__))

from biovault_agent import BioVaultAgent, BioVaultAgentError, AgentConfig


async def test_connection():
    """Test basic connection to the agent bridge."""
    async with BioVaultAgent() as agent:
        assert agent._ws is not None
        print("  Connection: OK")


async def test_discover():
    """Test API discovery endpoint."""
    async with BioVaultAgent() as agent:
        result = await agent.discover()
        assert "version" in result
        assert "name" in result
        assert result["name"] == "BioVault Desktop Agent API"
        print(f"  API Version: {result['version']}")


async def test_get_app_version():
    """Test getting app version."""
    async with BioVaultAgent() as agent:
        version = await agent.get_app_version()
        assert isinstance(version, str)
        assert len(version) > 0
        print(f"  App Version: {version}")


async def test_is_dev_mode():
    """Test dev mode check."""
    async with BioVaultAgent() as agent:
        is_dev = await agent.is_dev_mode()
        assert isinstance(is_dev, bool)
        print(f"  Dev Mode: {is_dev}")


async def test_check_is_onboarded():
    """Test onboarding status check."""
    async with BioVaultAgent() as agent:
        is_onboarded = await agent.check_is_onboarded()
        assert isinstance(is_onboarded, bool)
        print(f"  Onboarded: {is_onboarded}")


async def test_get_syftbox_state():
    """Test SyftBox state retrieval."""
    async with BioVaultAgent() as agent:
        state = await agent.get_syftbox_state()
        assert isinstance(state, dict)
        assert "running" in state
        print(f"  SyftBox Running: {state.get('running')}")


async def test_check_dependencies():
    """Test dependency checking."""
    async with BioVaultAgent() as agent:
        deps = await agent.check_dependencies()
        assert isinstance(deps, list)
        print(f"  Dependencies: {len(deps)} found")
        for dep in deps[:3]:  # Show first 3
            status = "installed" if dep.get("installed") else "missing"
            print(f"    - {dep.get('name')}: {status}")


async def test_get_projects():
    """Test project listing."""
    async with BioVaultAgent() as agent:
        projects = await agent.get_projects()
        assert isinstance(projects, list)
        print(f"  Projects: {len(projects)} found")


async def test_sql_list_tables():
    """Test SQL table listing."""
    async with BioVaultAgent() as agent:
        tables = await agent.sql_list_tables()
        assert isinstance(tables, list)
        print(f"  Database Tables: {len(tables)}")
        for table in tables[:5]:  # Show first 5
            print(f"    - {table}")


async def test_audit_log():
    """Test audit log functionality."""
    async with BioVaultAgent() as agent:
        # Get audit log
        log = await agent.get_audit_log(max_entries=5)
        assert isinstance(log, list)
        print(f"  Audit Log Entries: {len(log)}")

        # Verify recent entry has expected fields
        if log:
            entry = log[-1]
            assert "timestamp" in entry
            assert "cmd" in entry
            assert "success" in entry


async def test_invalid_command():
    """Test that invalid commands return errors."""
    async with BioVaultAgent() as agent:
        try:
            await agent.invoke("this_command_does_not_exist")
            assert False, "Should have raised an error"
        except BioVaultAgentError as e:
            assert "Unhandled command" in str(e)
            print("  Invalid command handling: OK")


async def test_auth_with_wrong_token():
    """Test authentication rejection with wrong token."""
    # Only run if token auth is configured
    if not os.environ.get("AGENT_BRIDGE_TOKEN"):
        print("  Skipping (no token configured)")
        return

    config = AgentConfig(token="wrong-token")
    async with BioVaultAgent(config) as agent:
        try:
            await agent.get_app_version()
            assert False, "Should have been rejected"
        except BioVaultAgentError as e:
            assert "Authentication failed" in str(e)
            print("  Auth rejection: OK")


async def run_all_tests():
    """Run all tests and report results."""
    tests = [
        ("Connection", test_connection),
        ("API Discovery", test_discover),
        ("Get App Version", test_get_app_version),
        ("Dev Mode Check", test_is_dev_mode),
        ("Onboarding Status", test_check_is_onboarded),
        ("SyftBox State", test_get_syftbox_state),
        ("Dependencies", test_check_dependencies),
        ("Projects", test_get_projects),
        ("SQL Tables", test_sql_list_tables),
        ("Audit Log", test_audit_log),
        ("Invalid Command", test_invalid_command),
        ("Auth Rejection", test_auth_with_wrong_token),
    ]

    passed = 0
    failed = 0
    skipped = 0

    print("\nBioVault Agent Client Tests")
    print("=" * 50)

    for name, test_fn in tests:
        print(f"\nTest: {name}")
        try:
            await test_fn()
            passed += 1
            print(f"  Result: PASSED")
        except AssertionError as e:
            failed += 1
            print(f"  Result: FAILED - {e}")
        except BioVaultAgentError as e:
            failed += 1
            print(f"  Result: FAILED - Agent error: {e}")
        except Exception as e:
            if "Skipping" in str(e) or "no token" in str(e):
                skipped += 1
                print(f"  Result: SKIPPED")
            else:
                failed += 1
                print(f"  Result: ERROR - {type(e).__name__}: {e}")

    print("\n" + "=" * 50)
    print(f"Results: {passed} passed, {failed} failed, {skipped} skipped")
    print("=" * 50)

    return failed == 0


# Pytest integration
import pytest

@pytest.fixture
def event_loop():
    loop = asyncio.new_event_loop()
    yield loop
    loop.close()


@pytest.mark.asyncio
async def test_connection_pytest():
    await test_connection()


@pytest.mark.asyncio
async def test_discover_pytest():
    await test_discover()


@pytest.mark.asyncio
async def test_get_app_version_pytest():
    await test_get_app_version()


@pytest.mark.asyncio
async def test_is_dev_mode_pytest():
    await test_is_dev_mode()


@pytest.mark.asyncio
async def test_check_is_onboarded_pytest():
    await test_check_is_onboarded()


@pytest.mark.asyncio
async def test_get_syftbox_state_pytest():
    await test_get_syftbox_state()


@pytest.mark.asyncio
async def test_check_dependencies_pytest():
    await test_check_dependencies()


@pytest.mark.asyncio
async def test_get_projects_pytest():
    await test_get_projects()


@pytest.mark.asyncio
async def test_sql_list_tables_pytest():
    await test_sql_list_tables()


@pytest.mark.asyncio
async def test_audit_log_pytest():
    await test_audit_log()


@pytest.mark.asyncio
async def test_invalid_command_pytest():
    await test_invalid_command()


if __name__ == "__main__":
    # Run tests directly
    success = asyncio.run(run_all_tests())
    sys.exit(0 if success else 1)
